ctapipe/tools/stage1.py

```diff
@@ -0,0 +1,868 @@
+        output HDF5 file
+    obs_id: int
+        observation ID
+    subarray:
+        SubarrayDescription to get metadata from
+    writer: HDF5TableWriter
+        output
+    """
+    import uuid
+
+    activity = PROV.current_activity.provenance
+
+    metadata = {
+        "CTA REFERENCE VERSION ": "1",
+        "CTA CONTACT ORGANIZATION": "CTA Consortium",
+        "CTA CONTACT NAME": "K. Kosack",
+        "CTA CONTACT EMAIL": "karl.kosack@cea.fr",
+        "CTA PRODUCT DESCRIPTION": "DL1 Event List",
+        "CTA PRODUCT CREATION_TIME": Time.now().iso,
+        "CTA PRODUCT ID": str(uuid.uuid4()),
+        "CTA PRODUCT DATA CATEGORY": "S",
+        "CTA PRODUCT DATA LEVEL": "DL1",
+        "CTA PRODUCT DATA TYPE": "Event",
+        "CTA PRODUCT DATA ASSOCIATION": "Subarray",
+        "CTA PRODUCT DATA MODEL NAME": "DL1/Event",
+        "CTA PRODUCT DATA MODEL VERSION": DL1_DATA_MODEL_VERSION,
+        "CTA PRODUCT DATA MODEL URL": "",
+        "CTA PRODUCT FORMAT": "hdf5",
+        "CTA PROCESS TYPE": "simulation",
+        "CTA PROCESS SUBTYPE": "",
+        "CTA PROCESS ID": str(obs_id),
+        "CTA ACTIVITY NAME": activity["activity_name"],
+        "CTA ACTIVITY TYPE": "software",
+        "CTA ACTIVITY ID": activity["activity_uuid"],
+        "CTA ACTIVITY START": activity["start"]["time_utc"],
+        "CTA ACTIVITY SOFTWARE NAME": "ctapipe",
+        "CTA ACTIVITY SOFTWARE VERSION": activity["system"]["ctapipe_version"],
+        "CTA INSTRUMENT SITE": "unknown",
+        "CTA INSTRUMENT CLASS": "subarray",
+        "CTA INSTRUMENT TYPE": "",
+        "CTA INSTRUMENT SUBTYPE": "",
+        "CTA INSTRUMENT VERSION": "",
+        "CTA INSTRUMENT ID": str(subarray),
+    }
+
+    for key, value in metadata.items():
+        writer._h5file.root._v_attrs[key] = value
+
+
+def morphology(geom, image_mask) -> MorphologyContainer:
+    """
+    Compute image morphology parameters
+    Parameters
+    ----------
+    geom: ctapipe.instrument.camera.CameraGeometry
+        camera description
+    image_mask: np.ndarray(bool)
+        image of pixels surviving cleaning (True=survives)
+    Returns
+    -------
+    MorphologyContainer:
+        parameters related to the morphology
+    """
+
+    num_islands, island_labels = number_of_islands(geom=geom, mask=image_mask)
+
+    return MorphologyContainer(num_pixels=image_mask.sum(), num_islands=num_islands)
+
+
+class IntensityContainer(Container):
+    """ Store statistics on the intensity distribution of images """
+
+    max = Field(np.nan, "value of pixel with maximum intensity")
+    min = Field(np.nan, "value of pixel with minimum intensity")
+    mean = Field(np.nan, "mean intensity")
+    std = Field(np.nan, "standard deviation of intensity")
+
+
+def intensity_statistics(image) -> IntensityContainer:
+    """ compute intensity statistics of an image  """
+    return IntensityContainer(
+        max=image.max(), min=image[image > 0].min(), mean=image.mean(), std=image.std()
```

Comment:
This > 0 is a bit surprising to me. shouldn't it be everywhere or nowhere?

So better call `image_statistics(image[image > 0])` instead and don't apply any masking here?
also include higher moments?

Reply:
To make the comment more explicit, 'intensity_statistics' is called on clean images, so to get the min, one clearly has to get zeroes out of the way... then one has to decide whether to keep them or no for the mean and std.
From the point of view of statistics, I agree with Max, better call `image_statistics(image[image > 0])`

Reply:
yes agreed, it's clearer. I don't like hidden cuts

Reply:
Just call it on the cleaned image or `image[cleaning]`